In [1]:
import io
import pathlib
import urllib.request
import zipfile

import numpy as np
import pandas as pd

# Background

Summary file 1:
https://www.census.gov/prod/cen2010/doc/sf1.pdf

Data File Segement contains population table 3 "P3" to population table 9 "P9". "P" prefixes denote census block population tables. "PCT" prefixes denote census tract tables. The target tagble is P5 (census doc page 184)

Census file 3:
    P3. RACE Universe: Total population (8)
    P4. HISPANIC OR LATINO ORIGIN Universe: Total population (3)
    P5. HISPANIC OR LATINO ORIGIN BY RACE Universe: Total population (17)

P5 data:
    Total: P0050001 03 9 
        Not Hispanic or Latino: P0050002 03 9 
            White alone P0050003 03 9 
            Black or African American alone P0050004 03 9 
            American Indian and Alaska Native alone P0050005 03 9 
            Asian alone P0050006 03 9 
            Native Hawaiian and Other Pacific Islander alone P0050007 03 9 
            Some Other Race alone P0050008 03 9 
            Two or More Races P0050009 03 9 
        Hispanic or Latino: P0050010 03 9 
            White alone P0050011 03 9 
            Black or African American alone P0050012 03 9 
            American Indian and Alaska Native alone P0050013 03 9 
            Asian alone P0050014 03 9 
            Native Hawaiian and Other Pacific Islander alone P0050015 03 9 
            Some Other Race alone P0050016 03 9 
            Two or More Races P0050017 03 9

Levels:
871 State-5-Digit ZIP Code Tabulation Area 
881 State-5-Digit ZIP Code Tabulation Area-County

All hispanic aggregated
Some other race alone (iterative prop)
Asian and API combined

In [4]:
# https://www.census.gov/prod/cen2010/doc/sf2.pdf
GEO_MAP_2010 = {
    'FILEID'  : (1  , 7  ),
    'STUSAB'  : (7  , 9  ),
    'SUMLEV'  : (9  , 12 ),
    'GEOCOMP' : (12 , 14 ),
    'CHARITER': (14 , 17 ),
    'CIFSN'   : (17 , 19 ),
    'LOGRECNO': (19 , 26 ),
    'REGION'  : (26 , 27 ),
    'DIVISION': (27 , 28 ),
    'STATE'   : (28 , 30 ),
    'COUNTY'  : (30 , 33 ),
    'COUNTYCC': (33 , 35 ),
    'COUNTYSC': (35 , 37 ),
    'COUSUB'  : (37 , 42 ),
    'COUSUBCC': (42 , 44 ),
    'COUSUBSC': (44 , 46 ),
    'PLACE'   : (46 , 51 ),
    'PLACECC' : (51 , 53 ),
    'PLACESC' : (53 , 55 ),
    'TRACT'   : (55 , 61 ),
    'BLKGRP'  : (61 , 62 ),
    'BLOCK'   : (62 , 66 ),
    'IUC'     : (66 , 68 ),
    'CONCIT'  : (68 , 73 ),
    'CONCITCC': (73 , 75 ),
    'CONCITSC': (75 , 77 ),
    'AIANHH'  : (77 , 81 ),
    'AIANHHFP': (81 , 86 ),
    'AIANHHCC': (86 , 88 ),
    'AIHHTLI' : (88 , 89 ),
    'AITSCE'  : (89 , 92 ),
    'AITS'    : (92 , 97 ),
    'AITSCC'  : (97 , 99 ),
    'TTRACT'  : (99 , 105),
    'TBLKGRP' : (105, 106),
    'ANRC'    : (106, 111),
    'ANRCCC'  : (111, 113),
    'CBSA'    : (113, 118),
    'CBSASC'  : (118, 120),
    'METDIV'  : (120, 125),
    'CSA'     : (125, 128),
    'NECTA'   : (128, 133),
    'NECTASC' : (133, 135),
    'NECTADIV': (135, 140),
    'CNECTA'  : (140, 143),
    'CBSAPCI' : (143, 144),
    'NECTAPCI': (144, 145),
    'UA'      : (145, 150),
    'UASC'    : (150, 152),
    'UATYPE'  : (152, 153),
    'UR'      : (153, 154),
    'CD'      : (154, 156),
    'SLDU'    : (156, 159),
    'SLDL'    : (159, 162),
    'VTD'     : (162, 168),
    'VTDI'    : (168, 169),
    'RESERVE2': (169, 172),
    'ZCTA5'   : (172, 177),
    'SUBMCD'  : (177, 182),
    'SUBMCDCC': (182, 184),
    'SDELM'   : (184, 189),
    'SDSEC'   : (189, 194),
    'SDUNI'   : (194, 199),
    'AREALAND': (119, 213),
    'AREAWATR': (213, 227),
    'NAME'    : (227, 317),
    'FUNCSTAT': (317, 318),
    'GCUNI'   : (318, 319),
    'POP100'  : (319, 328),
    'HU100'   : (328, 337),
    'INTPTLAT': (337, 348),
    'INTPTLON': (348, 360),
    'LSADC'   : (360, 362),
    'PARTFLAG': (362, 363),
    'RESERVE3': (363, 369),
    'UGA'     : (369, 374),
    'STATENS' : (374, 382),
    'COUNTYNS': (382, 390),
    'COUSUBNS': (390, 398),
    'PLACENS' : (398, 406),
    'CONCITNS': (406, 414),
    'AIANHHNS': (414, 422),
    'AITSNS'  : (422, 430),
    'ANRCNS'  : (430, 438),
    'SUBMCDNS': (438, 446),
    'CD113'   : (446, 448),
    'CD114'   : (448, 450),
    'CD115'   : (450, 452),
    'SLDU2'   : (452, 455),
    'SLDU3'   : (455, 458),
    'SLDU4'   : (458, 461),
    'SLDL2'   : (461, 464),
    'SLDL3'   : (464, 467),
    'SLDL4'   : (467, 470),
    'AIANHHSC': (470, 472),
    'CSASC'   : (472, 476),
    'CNECTASC': (474, 477),
    'MEMI'    : (476, 478),
    'NMEMI'   : (477, 478),
    'PUMA'    : (478, 483),
    'RESERVED': (483, 501),
}

# https://www.census.gov/prod/cen2010/doc/sf2.pdf
FILE_1_DATA_COLS = [
    'FILEID',
    'STUSAB',
    'CHARITER',
    'CIFSN',
    'LOGRECNO',
    'TOTAL_POPULATION',
]

STATES = {
    'AL': 'Alabama',
#    'AK': 'Alaska',
#    'AZ': 'Arizona',
#     'AR': 'Arkansas',
#     'CA': 'California',
#     'CO': 'Colorado',
#     'CT': 'Connecticut',
#     'DE': 'Delaware',
#     'DC': 'District_of_Columbia',
#     'FL': 'Florida',
#     'GA': 'Georgia',
#     'HI': 'Hawaii',
#     'ID': 'Idaho',
#     'IL': 'Illinois',
#     'IN': 'Indiana',
#     'IA': 'Iowa',
#     'KS': 'Kansas',
#     'KY': 'Kentucky',
#     'LA': 'Louisiana',
#     'ME': 'Maine',
#     'MD': 'Maryland',
#     'MA': 'Massachusetts',
#     'MI': 'Michigan',
#     'MN': 'Minnesota',
#     'MS': 'Mississippi',
#     'MO': 'Missouri',
#     'MT': 'Montana',
#     'NE': 'Nebraska',
#     'NV': 'Nevada',
#     'NH': 'New_Hampshire',
#     'NJ': 'New_Jersey',
#     'NM': 'New_Mexico',
#     'NY': 'New_York',
#     'NC': 'North_Carolina',
#     'ND': 'North_Dakota',
#     'OH': 'Ohio',
#     'OK': 'Oklahoma',
#     'OR': 'Oregon',
#     'PA': 'Pennsylvania',
#     'PR': 'Puerto_Rico',
#     'RI': 'Rhode_Island',
#     'SC': 'South_Carolina',
#     'SD': 'South_Dakota',
#     'TN': 'Tennessee',
#     'TX': 'Texas',
#     'UT': 'Utah',
#     'VT': 'Vermont',
#     'VA': 'Virginia',
#     'WA': 'Washington',
#     'WV': 'West_Virginia',
#     'WI': 'Wisconsin',
#     'WY': 'Wyoming',    
}

URL_TEMPLATE_ZIP = 'https://www2.census.gov/census_2010/04-Summary_File_1/{state}/{state_abbrev}2010.sf1.zip'

# Create URLS
urls = {
    code.lower(): URL_TEMPLATE_ZIP.format(state_abbrev=code.lower(), state=name)
    for code, name
    in STATES.items()
}

In [5]:
def download_file(url):
    print('.', end='')
    with urllib.request.urlopen(url) as f:
        data = io.BytesIO(f.read())
        return data


data = {
    code: download_file(url)
    for code, url
    in urls.items()
}

.

KeyboardInterrupt: 

In [ ]:
# PROCESS
data

In [3]:
# https://www2.census.gov/census_2000/
import os
os.listdir('C:/Users/theon/OneDrive/Desktop')

['2019.10.23.ansari_poa.docx',
 '2019.10.23.ansari_poa.pdf',
 'B2FE22A.tmp',
 'desktop.ini',
 'escrow_statement.pdf',
 'holding',
 'junk.html',
 'Microsoft Edge.lnk',
 'mo2010.sf1',
 'mo2010.sf1.zip',
 'NaunheimResume2019.docx',
 'otherstuff',
 'temp',
 'TODO.txt',
 'Uplay.lnk',
 'us2010.sf1.zip']

In [14]:
with zipfile.ZipFile('C:/Users/theon/OneDrive/Desktop/mo2010.sf1.zip') as zf:
    # Filter out everything except the ZipInfo for csv we want
    target = zf.filelist[0]
    # Read that CSV into BytesIO object
    raw_data = io.BytesIO(zf.read(target))
    geo = pd.read_fwf(
        raw_data, 
        header=None,
        # Subtract 1
        colspecs=[
            (tuple_[0] - 1, tuple_[1] - 1)
            for tuple_
            in GEO_MAP_2010.values()
        ],
        dtype=str
)


geo.columns = tuple(GEO_MAP_2010.keys())
geo = geo.loc[geo.SUMLEV == '871']    


geo_orig = geo.copy()
geo = geo[['STUSAB', 'LOGRECNO', 'ZCTA5']].dropna(subset=['ZCTA5'])

geo_orig

FILEID STUSAB SUMLEV GEOCOMP CHARITER CIFSN LOGRECNO REGION DIVISION  \
389482  SF1ST     MO    871      00      000   NaN  0389483      2        4   
389484  SF1ST     MO    871      00      000   NaN  0389485      2        4   
389486  SF1ST     MO    871      00      000   NaN  0389487      2        4   
389488  SF1ST     MO    871      00      000   NaN  0389489      2        4   
389490  SF1ST     MO    871      00      000   NaN  0389491      2        4   
389492  SF1ST     MO    871      00      000   NaN  0389493      2        4   
389494  SF1ST     MO    871      00      000   NaN  0389495      2        4   
389496  SF1ST     MO    871      00      000   NaN  0389497      2        4   
389498  SF1ST     MO    871      00      000   NaN  0389499      2        4   
389500  SF1ST     MO    871      00      000   NaN  0389501      2        4   
389503  SF1ST     MO    871      00      000   NaN  0389504      2        4   
389506  SF1ST     MO    871      00      000   NaN  0389507      2        4   
389508  SF1ST     MO    871      00      000   NaN  0389509      2        4   
389510  SF1ST     MO    871      00      000   NaN  0389511      2        4   
389512  SF1ST     MO    871      00      000   NaN  0389513      2        4   
389515  SF1ST     MO    871      00      000   NaN  0389516      2        4   
389517  SF1ST     MO    871      00      000   NaN  0389518      2        4   
389519  SF1ST     MO    871      00      000   NaN  0389520      2        4   
389522  SF1ST     MO    871      00      000   NaN  0389523      2        4   
389525  SF1ST     MO    871      00      000   NaN  0389526      2        4   
389528  SF1ST     MO    871      00      000   NaN  0389529      2        4   
389530  SF1ST     MO    871      00      000   NaN  0389531      2        4   
389532  SF1ST     MO    871      00      000   NaN  0389533      2        4   
389534  SF1ST     MO    871      00      000   NaN  0389535      2        4   
389536  SF1ST     MO    871      00      000   NaN  0389537      2        4   
389539  SF1ST     MO    871      00      000   NaN  0389540      2        4   
389541  SF1ST     MO    871      00      000   NaN  0389542      2        4   
389543  SF1ST     MO    871      00      000   NaN  0389544      2        4   
389545  SF1ST     MO    871      00      000   NaN  0389546      2        4   
389547  SF1ST     MO    871      00      000   NaN  0389548      2        4   
...       ...    ...    ...     ...      ...   ...      ...    ...      ...   
391987  SF1ST     MO    871      00      000   NaN  0391988      2        4   
391989  SF1ST     MO    871      00      000   NaN  0391990      2        4   
391992  SF1ST     MO    871      00      000   NaN  0391993      2        4   
391996  SF1ST     MO    871      00      000   NaN  0391997      2        4   
391998  SF1ST     MO    871      00      000   NaN  0391999      2        4   
392001  SF1ST     MO    871      00      000   NaN  0392002      2        4   
392003  SF1ST     MO    871      00      000   NaN  0392004      2        4   
392006  SF1ST     MO    871      00      000   NaN  0392007      2        4   
392010  SF1ST     MO    871      00      000   NaN  0392011      2        4   
392012  SF1ST     MO    871      00      000   NaN  0392013      2        4   
392014  SF1ST     MO    871      00      000   NaN  0392015      2        4   
392017  SF1ST     MO    871      00      000   NaN  0392018      2        4   
392020  SF1ST     MO    871      00      000   NaN  0392021      2        4   
392022  SF1ST     MO    871      00      000   NaN  0392023      2        4   
392024  SF1ST     MO    871      00      000   NaN  0392025      2        4   
392028  SF1ST     MO    871      00      000   NaN  0392029      2        4   
392031  SF1ST     MO    871      00      000   NaN  0392032      2        4   
392033  SF1ST     MO    871      00      000   NaN  0392034      2        4   
392036  SF1ST     MO    871      00      000   NaN  03920

In [15]:
with zipfile.ZipFile('C:/Users/theon/OneDrive/Desktop/mo2010.sf1.zip') as zf:
    # Filter out everything except the ZipInfo for csv we want
    target = zf.filelist[3]
    print(target.filename)
    # Read that CSV into BytesIO object
    raw_data = io.BytesIO(zf.read(target))
    data = pd.read_csv(
        raw_data, 
        header=None,
        dtype=str
    )
    pop = data[[1, 4, 18, 19, 20, 21, 22, 23, 24, 25]]
    pop.columns = [
        'STUSAB',
        'LOGRECNO',
        'white',
        'black',
        'native',
        'asian',
        'pi',
        'other',
        'multiple',
        'hispanic',
    ]
    
pop.head()

mo000032010.sf1


STUSAB LOGRECNO    white   black native  asian    pi other multiple hispanic
0     MO  0000001  4850748  687149  24062  97221  5763  5372   106142   212470
1     MO  0000002        0       0      0      0     0     0        0        0
2     MO  0000003        0       0      0      0     0     0        0        0
3     MO  0000004        0       0      0      0     0     0        0        0
4     MO  0000005        0       0      0      0     0     0        0        0

In [16]:
merged = geo.merge(pop)
merged = merged.set_index('ZCTA5')
merged = merged.sort_index()
merged

STUSAB LOGRECNO  white  black native asian  pi other multiple hispanic
ZCTA5                                                                       
51640     MO  0389483     11      0      0     0   0     0        0        0
52542     MO  0389485     40      0      0     0   0     0        0        0
52573     MO  0389487     25      0      0     0   0     0        0        0
52626     MO  0389489    101      0      0     0   0     0        0        0
63005     MO  0389491  15628    348     25  1061   5    15      220      451
63010     MO  0389493  33697    196     66   307   5    15      392      761
63011     MO  0389495  32319    811     52  1803   3    37      494      828
63012     MO  0389497   9615     44     26    30   2     5       92      159
63013     MO  0389499   1567      4      0     1   0     0        9        8
63014     MO  0389501    770      2      0     0   0     1        2        4
63015     MO  0389504   2012      6      8     5   0     1       26       18
63016     MO  0389507   7314     14     37    26   0     7       73      111
63017     MO  0389509  35051   1134     54  3536  11    41      509     1134
63019     MO  0389511   4098    168     16    28   1     9       84       45
63020     MO  0389513  19968    165     56    64   4    26      235      196
63021     MO  0389516  48149   1395     78  3910  16    51      959     1703
63023     MO  0389518   5648     15     19    17   0     1       65       66
63025     MO  0389520  13036     84     33   210  10     8      181      292
63026     MO  0389523  41278    342     99   940  20    20      559      866
63028     MO  0389526  24996    471     67   156   9    15      396      303
63030     MO  0389529    400      1      0     0   0     0        1        5
63031     MO  0389531  34092  11313     98   515  22    73     1043     1056
63033     MO  0389533  15199  25992     61   260   7    65      829      587
63034     MO  0389535   7313   9533     23   327  11    31      279      244
63036     MO  0389537   1034      7      6     3   2     0        4        8
63037     MO  0389540   3169      3     12    10   0     2       25       20
63038     MO  0389542   6373    103     13   154   3     6       88      174
63039     MO  0389544    691      0      1    18   0     0        9        7
63040     MO  0389546   7687    138      7   387   0     5      125      163
63041     MO  0389548    511      2      0     1   5     0       14        1
...      ...      ...    ...    ...    ...   ...  ..   ...      ...      ...
65768     MO  0391988    471      0      4     0   0     0       10        3
65769     MO  0391990   2528      5     28    13   1     0       32      484
65770     MO  0391993   2881      3     27     8   0     3       41       39
65771     MO  0391997   1309      5     11     7   0     1       29       36
65772     MO  0391999   2045      6     11    24   2     0       39       31
65773     MO  0392002    680      1      1     1   0     0       13       15
65774     MO  0392004    926      1      8     1   1     0       15        1
65775     MO  0392007  23641    121    153   133  12     9      356      483
65777     MO  0392011    306      0      3     0   0     0       13        6
65778     MO  0392013    514      1      6     0   0     0        1        2
65779     MO  0392015   2116      7     15     2   7     0       39       21
65781     MO  0392018   8179     58     32     8   5     2      126      133
65783     MO  0392021    207      3      1     0   0     0       13        1
65784     MO  0392023     33      0      0     0   0     0        0        1
65785     MO  0392025   5055      5     26    16   3     0       96       58
65786     MO  0392029   2128      1     24     4   0     7       37       49
65787     MO  0392032   1790      0      6    13   0     0       28       27
65788     MO  0392034    410      0      0     1   0     0        9        2
65789     MO  0392037   1911      5     22    15   0     2       26       33
65

In [59]:
jg2 = jgeo.merge(jdata)
jg2['RACE'] = 'White'
jg2

STUSAB LOGRECNO  ZCTA5 TOTAL_POPULATION   RACE
0          MO  0000025  63501                0  White
1          MO  0000026  63501                0  White
2          MO  0000027  63501                0  White
3          MO  0000028  63501                0  White
4          MO  0000029  63501               31  White
5          MO  0000030  63501                2  White
6          MO  0000031  63501                1  White
7          MO  0000033  63501                0  White
8          MO  0000034  63501                0  White
9          MO  0000035  63501               19  White
10         MO  0000036  63501                0  White
11         MO  0000037  63501                5  White
12         MO  0000038  63501               16  White
13         MO  0000039  63501                0  White
14         MO  0000040  63501                0  White
15         MO  0000041  63501                0  White
16         MO  0000042  63501                0  White
17         MO  0000045  63501                0  White
18         MO  0000046  63501                0  White
19         MO  0000047  63501                0  White
20         MO  0000048  63501                0  White
21         MO  0000051  63501               10  White
22         MO  0000052  63501                0  White
23         MO  0000053  63501                0  White
24         MO  0000054  63501                0  White
25         MO  0000055  63501                0  White
26         MO  0000056  63501                0  White
27         MO  0000057  63501                0  White
28         MO  0000058  63501                0  White
29         MO  0000059  63501                0  White
...       ...      ...    ...              ...    ...
346116     MO  0392034  65788              422  White
346117     MO  0392035  65788              406  White
346118     MO  0392036  65788               16  White
346119     MO  0392037  65789             2014  White
346120     MO  0392038  65789             2014  White
346121     MO  0392039  65790              739  White
346122     MO  0392040  65790              643  White
346123     MO  0392041  65790               96  White
346124     MO  0392042  65791             4585  White
346125     MO  0392043  65791             4585  White
346126     MO  0392044  65793             6337  White
346127     MO  0392045  65793              244  White
346128     MO  0392046  65793             5439  White
346129     MO  0392047  65793              654  White
346130     MO  0392048  65802            44270  White
346131     MO  0392049  65802            44270  White
346132     MO  0392050  65803            40902  White
346133     MO  0392051  65803            40902  White
346134     MO  0392052  65804            36801  White
346135     MO  0392053  65804            36801  White
346136     MO  0392054  65806            10586  White
346137     MO  0392055  65806            10586  White
346138     MO  0392056  65807            54952  White
346139     MO  0392057  65807            54952  White
346140     MO  0392058  65809            11257  White
346141     MO  0392059  65809            11257  White
346142     MO  0392060  65810            20984  White
346143     MO  0392061  65810            20984  White
346144     MO  0392062  72644               15  White
346145     MO  0392063  72644               15  White

[346146 rows x 5 columns]

PCT1: Total population
PCT0010001 

DATA FILE SEGMENT 1, CELL 1 PCT1

Iterations

001: Total Population
400: Hispanic of Latino of Any Race
450: All Not Hispanic
451: White Not Hispanic
453: Black Not Hispanic
455: Native Not Hispanic
457: Asian Not Hispanic
459: Pacific Islander Not Hispanic
463: Multiple

->

white
black
asian
native
multiple
hispanic

# Background (links good as of 2019-12-07)

This fetches data from the following zipfiles:

* https://www2.census.gov/topics/genealogy/2000surnames/names.zip
* https://www2.census.gov/topics/genealogy/2010surnames/names.zip

These zipfiles contains CSVs with race data based on the surnames. The script creates dataframes from the CSVs in these zipfiles. It then cleans them and imputes data to remove anonymization in accordance with the rules laid out below.

Information about the contents of these files may be found here:

* http://www2.census.gov/topics/genealogy/2000surnames/surnames.pdf?#
* https://www2.census.gov/topics/genealogy/2010surnames/surnames.pdf

It then the data for use in Surgeo calculations.

# Constants

In [2]:
CENSUS_URL_2000 = 'https://www2.census.gov/topics/genealogy/2000surnames/names.zip'

CENSUS_URL_2010 = 'https://www2.census.gov/topics/genealogy/2010surnames/names.zip'

CENSUS_SURNAME_COLUMNS = [
    'name',
    'rank',
    'count',
    'proportion',
    'cum_proportion',
    'white',
    'black',
    'asian',
    'native',
    'multiple',
    'hispanic',
]

TARGET_SURNAME_COLUMNS = [
    'name',
    'white',
    'black',
    'asian',
    'native',
    'multiple',
    'hispanic',
]

## Download Surname Data

In [3]:
def url_to_df(url):
    '''Takes the URL of a Census zip file and converts to DF
    
    Note: it appears the Census webservers rate limit this so it
    may take some time.
    
    '''
    # Download zipfile from census URL
    with urllib.request.urlopen(url) as response:
        # Write file into BytesIO object
        zip_data = io.BytesIO(response.read())
        # Open zip data as zipfile
        with zipfile.ZipFile(zip_data) as zf:
            # Filter out everything except the ZipInfo for csv we want
            csv_info = [file for file in zf.filelist if '.csv' in file.filename][0]
            # Read that CSV into BytesIO object
            raw_data = io.BytesIO(zf.read(csv_info))
            # Create dataframe with only suppressed '(S)' converted to NA
            df = pd.read_csv(raw_data, na_values='(S)', keep_default_na=False)
            return df

In [4]:
df_2000 = url_to_df(CENSUS_URL_2000)
df_2010 = url_to_df(CENSUS_URL_2010)

In [5]:
df_2010.tail()

name    rank     count  prop100k  cum_prop100k  pctwhite  \
162249        DIETZMANN  160975       100      0.03      90062.93     96.00   
162250            DOKAS  160975       100      0.03      90062.96     94.00   
162251           DONLEA  160975       100      0.03      90062.99     94.00   
162252          DORIOTT  160975       100      0.03      90063.03     89.00   
162253  ALL OTHER NAMES       0  29312001   9936.97       9936.97     66.65   

        pctblack  pctapi  pctaian  pct2prace  pcthispanic  
162249      0.00    0.00      NaN       0.00          NaN  
162250       NaN    0.00     0.00        NaN          NaN  
162251      0.00    0.00     0.00       0.00         6.00  
162252      0.00     NaN     0.00       5.00          NaN  
162253      8.53    7.97     0.86       2.32        13.67

# Clean Data

In [6]:
def clean_df(df):
    '''Change column names, set index, and convert percentages'''
    # Change names
    df.columns = CENSUS_SURNAME_COLUMNS
    # Filter columns
    df = df[TARGET_SURNAME_COLUMNS]
    # Set index to name
    df = df.set_index('name')
    # Sort index
    df = df.sort_index()
    # Convert percentages to 0 to 1 numbers
    df = df / 100
    return df

In [7]:
df_2000 = clean_df(df_2000)
df_2010 = clean_df(df_2010)

In [8]:
df_2010.tail()

white  black  asian  native  multiple  hispanic
name                                                       
ZYSK       0.9873    0.0    NaN     NaN    0.0000       NaN
ZYSKOWSKI  0.9655    NaN    NaN     0.0    0.0164    0.0127
ZYSMAN     0.9457    NaN    0.0     0.0       NaN       NaN
ZYWICKI    0.9552    NaN    NaN     0.0    0.0125    0.0233
ZYWIEC     0.9829    NaN    0.0     0.0    0.0000       NaN

# Unsuppress / impute anonymized data

In [9]:
def unsuppress_row(row):
    '''Apply function to desuppress data on row-basis
    
    If a percentage falls beneath a certain threshold, the
    US Census quasi-anonymizes it by supressing. To impute
    new values, we get the outstanding allocated percentage
    and divy it up among the suppressed fields.
    
    '''
    # Check if row has NA values
    if row.isna().sum() > 0:
        # Get count of NA values
        na_count = row.isna().sum()
        # Get total of percentages
        row_sum = row.sum()
        # Get unallocated percentage and divide by count of NA
        na_value = (1 - row_sum) / na_count
        # Fill NA values with that row value.
        reconstituted_row = row.fillna(na_value)
        # Round if necessary
        return reconstituted_row.round(4)
    else:
        # If there's no NA, there's no need to impute
        return row

In [10]:
# Get rows with NaNs
target_2000 = df_2000.isna().any(axis=1)
# Run this inefficient operation on rows with NaNs only
df_2000.loc[target_2000] = (
    df_2000.loc[target_2000].apply(unsuppress_row, axis=1)
)

# Get rows with NaNs
target_2010 = df_2010.isna().any(axis=1)
# Run this inefficient operation on rows with NaNs only
df_2010.loc[target_2010] = (
    df_2010.loc[target_2010].apply(unsuppress_row, axis=1)
)

In [11]:
df_2010.tail()

white   black   asian  native  multiple  hispanic
name                                                         
ZYSK       0.9873  0.0000  0.0042  0.0042    0.0000    0.0042
ZYSKOWSKI  0.9655  0.0027  0.0027  0.0000    0.0164    0.0127
ZYSMAN     0.9457  0.0181  0.0000  0.0000    0.0181    0.0181
ZYWICKI    0.9552  0.0045  0.0045  0.0000    0.0125    0.0233
ZYWIEC     0.9829  0.0085  0.0000  0.0000    0.0000    0.0085

# Write data to module as CSV

In [12]:
current_directory = pathlib.Path().cwd()
project_directory = current_directory.parents[0]
data_directory    = project_directory / 'surgeo' / 'data'
path_2000         = data_directory / 'surnames_2000.csv'
path_2010         = data_directory / 'surnames_2010.csv'
df_2000.to_csv(path_2000)
df_2010.to_csv(path_2010)